File: 2023/typescript/src/day01.ts
import * as fs from "fs";
import _ from "lodash";

day01();

function day01(): void {
  const lines: string[] = fs
    .readFileSync("data/day01-input.txt", { encoding: "utf8" })
    .trim()
    .split("\n");
  console.log("Part one:", _.sum(lines.map(findCalibrationNumber)));
  // Part two is off by just a bit (5) and I can't figure out why.
  console.log("Part two:", _.sum(lines.map(spellingsToDigits).map(findCalibrationNumber)));
}

function findCalibrationNumber(line: string): number {
  const digits: string[] = [...line].filter((c) => c >= "1" && c <= "9");
  return parseInt(digits[0] + _.last(digits));
}

function spellingsToDigits(line: string): string {
  if (line.length === 0) return line;
  const spellings = ["one", "two", "three", "four", "five", "six", "seven", "eight", "nine"];
  for (let i = 0; i < spellings.length; i++) {
    const spelling = spellings[i];
    if (line.startsWith(spelling)) {
      return (i + 1).toString() + spellingsToDigits(line.slice(spelling.length));
    }
  }
  return line[0] + spellingsToDigits(line.slice(1));
}

File: 2023/haskell/src/Day01.hs
module Day01 where

import Control.Arrow (first)
import Data.Function ((&))
import Data.List (find, isPrefixOf)
import Paths_aoc2023 (getDataFileName)

day01 :: IO ()
day01 = do
  inputLines <- lines <$> (getDataFileName "day01-input.txt" >>= readFile)
  putStrLn $ "Part one: " <> show (sum $ findCalibrationValue numberSpellings <$> inputLines)
  putStrLn $ "Part two: " <> show (sum $ findCalibrationValue (numberSpellings <> wordSpellings) <$> inputLines)
  where
    numberSpellings = zip (show <$> [0 :: Int .. 9]) [0 ..]
    wordSpellings = zip ["one", "two", "three", "four", "five", "six", "seven", "eight", "nine"] [1 ..]

type Line = String

type Spellings = [(String, Int)]

findCalibrationValue :: Spellings -> Line -> Int
findCalibrationValue spellings line = 10 * findFirstDigit spellings line + findLastDigit spellings line

findFirstDigit :: Spellings -> Line -> Int
findFirstDigit _ [] = error "No digit found -> this should never happen!"
findFirstDigit spellings s@(_ : cs) =
  find ((`isPrefixOf` s) . fst) spellings
    & maybe (findFirstDigit spellings cs) snd

findLastDigit :: Spellings -> Line -> Int
findLastDigit spellings = findFirstDigit (first reverse <$> spellings) . reverse
